src/node_contextify.cc

```diff
@@ -441,40 +373,111 @@ class ContextifyContext {
-      maybe_prop_attr =
-          ctx->global_proxy()->GetRealNamedPropertyAttributes(context,
-                                                              property);
-    }
+    Local<String> key = property->ToString(isolate);
+    Local<Object> sandbox = ctx->sandbox();
 
-    if (maybe_prop_attr.IsJust()) {
-      PropertyAttribute prop_attr = maybe_prop_attr.FromJust();
-      args.GetReturnValue().Set(prop_attr);
+    Maybe<bool> has = sandbox->HasOwnProperty(context, key);
+    Local<Value> descriptor_intercepted = has.IsNothing() ?
+        ctx->global_proxy()->GetOwnPropertyDescriptor(context, key)
+            .ToLocalChecked() :
+        sandbox->GetOwnPropertyDescriptor(context, key).ToLocalChecked();
+
+    // If the property had already been defined on the sandbox or global object,
+    // we intercept, query and return it.
+    // Else, there is no need to intercept with first time definitions.
+    if (!descriptor_intercepted->IsUndefined()) {
+      info.GetReturnValue().Set(descriptor_intercepted);
     }
   }
 
 
+  static void GlobalPropertyDefinerCallback(
+      Local<Name> property,
+      const PropertyDescriptor& desc,
+      const PropertyCallbackInfo<Value>& info) {
+    ContextifyContext* ctx;
+    ASSIGN_OR_RETURN_UNWRAP(&ctx, info.Data().As<Object>());
+
+    // Still initializing
+    if (ctx->context_.IsEmpty())
+      return;
+
+    Local<Context> context = ctx->context();
+    v8::Isolate* isolate = context->GetIsolate();
+    HandleScope scope(isolate);
+
+    auto attributes = PropertyAttribute::None;
+    bool is_declared =
+        ctx->global_proxy()->GetRealNamedPropertyAttributes(ctx->context(),
+                                                            property)
+            .To(&attributes);
+    bool read_only =
+        static_cast<int>(attributes) &
+            static_cast<int>(PropertyAttribute::ReadOnly);
+
+    // If the property is set on the global as read_only, don't change it on
+    // the global or sandbox.
+    if (is_declared && read_only)
+      return;
+
+    Local<Object> sandbox = ctx->sandbox();
+
+    auto define_prop_on_sandbox =
+        [&] (PropertyDescriptor* desc_for_sandbox) {
+          if (desc.has_enumerable()) {
+            desc_for_sandbox->set_enumerable(desc.enumerable());
+          }
+          if (desc.has_configurable()) {
+            desc_for_sandbox->set_configurable(desc.configurable());
+          }
+          // Set the property on the sandbox.
+          sandbox->DefineProperty(context, property, *desc_for_sandbox);
+        };
+
+    if (desc.has_get() || desc.has_set()) {
+      PropertyDescriptor desc_for_sandbox(
+          desc.has_get() ? desc.get() : v8::Undefined(isolate).As<Value>(),
+          desc.has_set() ? desc.set() : v8::Undefined(isolate).As<Value>());
+
+      define_prop_on_sandbox(&desc_for_sandbox);
+    } else {
+      Local<Value> value =
+          desc.has_value() ? desc.value() :
+          v8::Undefined(isolate).As<Value>();
```

Comment:
This fits on one line (well, two, if you count the LHS.)

Reply:
Done.